chapter6/model-action-threshold/experiment.py

```
@@ -0,0 +1,718 @@
+
+def changed_files(before: dict[str, str], repo: Path) -> list[str]:
+    after = snapshot(repo)
+    return sorted(
+        path for path in set(before) | set(after) if before.get(path) != after.get(path)
+    )
+
+
+def run_test_command(repo: Path, command: list[str], timeout: int = 30) -> dict[str, Any]:
+    started = time.monotonic()
+    try:
+        completed = subprocess.run(
```

Ruff (0.16.0)

[error] `subprocess` call: check for execution of untrusted input

(S603)